[extension/bearertokenauth] feat add telemetry for tracking token usage

brdotv2 opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
Adding internal metrics to the extension to be able to measure token usage.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/45047

<!--Describe what testing was performed and which tests were added.-->
#### Testing

Tested that the new metrics are correctly registered and exposed.

<!--Describe the documentation added.-->
#### Documentation

```
attributes:
  auth.status:
    description: Authentication status
    type: string
    enum: [success, failure]
  auth.type:
    description: Type of authentication
    type: string
    enum: [client, server]
  token.index:
    description: Index of the token in the configuration (0-based, -1 when token.matched is false)
    type: int
  token.matched:
    description: Whether the provided token matches a configured token
    type: bool

telemetry:
  metrics:
    bearer_token_auth_usage:
      enabled: true
      description: Number of authentication attempts with bearer tokens
      unit: "{attempt}"
      stability:
        level: development
      sum:
        value_type: int
        monotonic: true
      attributes: [auth.type, auth.status, token.index, token.matched]
```

brdotv2 committed:
barertokenauth - add telemetry for tracking token usage

brdotv2 committed:
add changelog entry

brdotv2 commented:
/label -stale

dehaansa commented on extension/bearertokenauthextension/bearertokenauth.go:
```diff
@@ -249,6 +285,19 @@ type bearerAuthRoundTripper struct {
 // RoundTrip modifies the original request and adds Bearer token Authorization headers. Incoming requests support multiple tokens, but outgoing requests only use one.
 func (interceptor *bearerAuthRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Track client authentication usage (always uses first token, index 0)
+	if interceptor.auth.telemetry != nil {
```
Can we capture failure here by deferring this logic until after the RoundTrip of the baseTransport?

dehaansa reviewed (commented):
Apologies for the slow review, I was out for the holidays and missed some github notifications.

brdotv2 committed:
PR suggestions

- fail fast if telemetry can't be initialized in newBearerTokenAuth
- defer capturing client call status til after the RoundTrip
